src/constraints/mod.rs

```diff
@@ -0,0 +1,184 @@
+    for bits in &all_nonnative_bits_le {
+        max_nonnative_bits = max_nonnative_bits.max(bits.len());
+    }
+
+    let mut lookup_table = Vec::<Vec<CF>>::new();
+    let mut cur = F::one();
+    for _ in 0..max_nonnative_bits {
+        let repr = AllocatedNonNativeFieldVar::<F, CF>::get_limbs_representations(
+            &cur,
+            OptimizationType::Constraints,
+        )?;
+        lookup_table.push(repr);
+        cur.double_in_place();
+    }
+
+    let params = get_params(
+        F::size_in_bits(),
+        CF::size_in_bits(),
+        OptimizationType::Constraints,
+    );
+
+    let mut output = Vec::with_capacity(all_nonnative_bits_le.len());
+    for nonnative_bits_le in all_nonnative_bits_le {
+        let mut val = vec![CF::zero(); params.num_limbs];
+        let mut lc = vec![LinearCombination::<CF>::zero(); params.num_limbs];
+
+        for (j, bit) in nonnative_bits_le.iter().enumerate() {
+            if bit.value().unwrap_or_default() {
+                for (k, val) in val.iter_mut().enumerate().take(params.num_limbs) {
+                    *val += &lookup_table[j][k];
+                }
+            }
+
+            #[allow(clippy::needless_range_loop)]
+            for k in 0..params.num_limbs {
+                lc[k] = &lc[k] + bit.lc() * lookup_table[j][k];
+            }
+        }
+
+        let mut limbs = Vec::new();
+        for k in 0..params.num_limbs {
+            let gadget =
+                AllocatedFp::new_witness(ark_relations::ns!(cs, "alloc"), || Ok(val[k])).unwrap();
+            lc[k] = lc[k].clone() - (CF::one(), gadget.variable);
+            cs.enforce_constraint(lc!(), lc!(), lc[k].clone()).unwrap();
+            limbs.push(FpVar::<CF>::from(gadget));
+        }
+
+        output.push(NonNativeFieldVar::<F, CF>::Var(
+            AllocatedNonNativeFieldVar::<F, CF> {
+                cs: cs.clone(),
+                limbs,
+                num_of_additions_over_normal_form: CF::zero(),
+                is_in_the_normal_form: true,
+                target_phantom: Default::default(),
+            },
+        ));
+    }
+
+    Ok(output)
+}
+
+/// The interface for a cryptographic sponge.
+/// A sponge can `absorb` or take in inputs and later `squeeze` or output bytes or field elements.
+/// The outputs are dependent on previous `absorb` and `squeeze` calls.
+/// For now, `CryptographicSpongeVar` only supports field-based sponge.
+pub trait CryptographicSpongeVar<CF: PrimeField, S: FieldBasedCryptographicSponge<CF>>:
+    Clone
+{
+    /// Parameters used by the sponge.
+    type Parameters;
+
+    /// Initialize a new instance of the sponge.
+    fn new(cs: ConstraintSystemRef<CF>, params: &Self::Parameters) -> Self;
+
+    /// Returns a ref to the underlying constraint system the sponge is operating in.
+    fn cs(&self) -> ConstraintSystemRef<CF>;
+
+    /// Absorb an input into the sponge.
+    fn absorb(&mut self, input: &impl AbsorbGadget<CF>) -> Result<(), SynthesisError>;
+
```

Comment:
So we don't have a way to absorb bytes other than converting them to field elements first?

Reply:
For now, yes. Currently constraints API only supports field-based sponge. 

If we plan to implement byte-based sponge constraints in the future, then `AbsorbGadget` should also have `absorb_bytes` API built-in. At that time we can directly absorb those bytes

Reply:
Yes, I think we should do that, as I'm pretty sure it would be useful for a number of projects.

Reply:
Great. Let me implement that (expect to finish those by weekend. I will send you an update once I've done)

Reply:
@Pratyush constraints API updated!